client_authentication.go

```diff
@@ -200,7 +203,10 @@ func (f *Fosite) AuthenticateClient(ctx context.Context, r *http.Request, form u
 	}
 
 	// Enforce client authentication
+	span, _ := opentracing.StartSpanFromContext(ctx, "hash_compare")
+	defer span.Finish()
 	if err := f.Hasher.Compare(client.GetHashedSecret(), []byte(clientSecret)); err != nil {
```

Comment:
This should all probably be moved into `Compare` itself so we can add additional useful tags - What do you think @aeneasr ? In order to do so, I will need to change the signature of the functions in [Hasher](https://github.com/ory/fosite/blob/master/hash.go#L28) and pass down context. For example:

```
type Hasher interface {
	// Compare compares data with a hash and returns an error
	// if the two do not match.
	Compare(ctx context.Context, hash, data []byte) error

	// Hash creates a hash from data or returns an error.
	Hash(ctx context.Context, data []byte) ([]byte, error)
}
```

This would give us the opportunity to add tags depending on the hasher used. For example, for bcrypt we would be able to add a tag for the [work factor](https://github.com/ory/fosite/blob/master/hash_bcrypt.go#L31) and the hashing algorithm that is being applied. This would benefit someone that is profiling their oauth svc.